File: src/main/scala/com/github/phisgr/gatling/grpc/action/GrpcCallActionBuilder.scala
package com.github.phisgr.gatling.grpc.action

import com.github.phisgr.gatling.grpc.check.GrpcCheck
import com.github.phisgr.gatling.grpc.request.CallAttributes
import io.gatling.core.action.Action
import io.gatling.core.session.Expression
import io.gatling.core.structure.ScenarioContext
import io.grpc.MethodDescriptor

case class GrpcCallActionBuilder[Req, Res](
  requestName: Expression[String],
  private[gatling] override val method: MethodDescriptor[Req, Res],
  payload: Expression[Req],
  private[gatling] override val callAttributes: CallAttributes = CallAttributes(),
  private[gatling] override val checks: List[GrpcCheck[Res]] = Nil,
  isSilent: Boolean = false
) extends UnaryResponseBuilder[GrpcCallActionBuilder[Req, Res], Req, Res] {

  override def build(ctx: ScenarioContext, next: Action): Action = new GrpcCallAction(this, ctx, next)

  override def check(checks: GrpcCheck[Res]*): GrpcCallActionBuilder[Req, Res] =
    copy(checks = this.checks ::: checks.toList)

  override private[gatling] def withCallAttributes(callAttributes: CallAttributes): GrpcCallActionBuilder[Req, Res] =
    copy(callAttributes = callAttributes)

  /**
   * If a request is silent, it will not be included in the final report.
   * On failure, it will not mark the session as failed.
   *
   * @return a new action builder with isSilent set to true
   */
  def silent: GrpcCallActionBuilder[Req, Res] =
    copy(isSilent = true)

}

File: kt/src/main/kotlin/com/github/phisgr/gatling/kt/grpc/action/GrpcCallActionBuilder.kt
package com.github.phisgr.gatling.kt.grpc.action

import com.github.phisgr.gatling.grpc.check.GrpcCheck
import com.github.phisgr.gatling.grpc.check.GrpcResponse
import com.github.phisgr.gatling.kt.grpc.internal.build
import com.github.phisgr.gatling.kt.grpc.request.CallDefinition
import io.gatling.javaapi.core.CheckBuilder
import com.github.phisgr.gatling.grpc.action.GrpcCallActionBuilder as GrpcCallActionBuilderS

class GrpcCallActionBuilder<Req, Res>(private val wrapped: GrpcCallActionBuilderS<Req, Res>) : CallDefinition<
    GrpcCallActionBuilder<Req, Res>,
    Req,
    Res,
    GrpcResponse<Res>,
    GrpcCallActionBuilderS<Req, Res>,
    GrpcCheck<Res>>() {
    override fun asScala(): GrpcCallActionBuilderS<Req, Res> = wrapped

    override fun buildCheck(check: CheckBuilder): GrpcCheck<Res> =
        check.build()

    override fun wrap(wrapped: GrpcCallActionBuilderS<Req, Res>): GrpcCallActionBuilder<Req, Res> =
        GrpcCallActionBuilder(wrapped)

}
